File: Number of Enclaves.cpp
class Solution {
public:
    int numEnclaves(vector<vector<int>>& grid) {
        int m = grid.size();
        int n = grid[0].size();
        vector<vector<bool>> visited(m, vector<bool>(n, false));
        
        // DFS or BFS on boundary cells
        for(int i = 0; i < m; i++) {
            if(grid[i][0] == 1 && !visited[i][0]) {
                dfs(grid, visited, i, 0);
            }
            if(grid[i][n-1] == 1 && !visited[i][n-1]) {
                dfs(grid, visited, i, n-1);
            }
        }
        for(int j = 0; j < n; j++) {
            if(grid[0][j] == 1 && !visited[0][j]) {
                dfs(grid, visited, 0, j);
            }
            if(grid[m-1][j] == 1 && !visited[m-1][j]) {
                dfs(grid, visited, m-1, j);
            }
        }
        
        // Count land cells that are not connected to the boundary
        int count = 0;
        for(int i = 1; i < m-1; i++) {
            for(int j = 1; j < n-1; j++) {
                if(grid[i][j] == 1 && !visited[i][j]) {
                    count++;
                }
            }
        }
        return count;
    }
    
    void dfs(vector<vector<int>>& grid, vector<vector<bool>>& visited, int i, int j) {
        visited[i][j] = true;
        int m = grid.size();
        int n = grid[0].size();
        int dx[] = {0, 0, 1, -1};
        int dy[] = {1, -1, 0, 0};
        for(int k = 0; k < 4; k++) {
            int x = i + dx[k];
            int y = j + dy[k];
            if(x >= 0 && x < m && y >= 0 && y < n && grid[x][y] == 1 && !visited[x][y]) {
                dfs(grid, visited, x, y);
            }
        }
    }
};

File: Number of Enclaves.java
class Pair{
    int row;
    int col;
    Pair(int row, int col){
        this.row = row;
        this.col = col;
    }
}
class Solution {
    public void dfs(int row, int col , int [][] grid, int [] lands){
        int N = grid.length;
        int M = grid[0].length;
        if(row < 0 || row == N || col< 0 || col == M || grid[row][col] == 0) return ;
        int [] delRow = {-1, 0, 1, 0};
        int [] delCol = {0, 1, 0, -1};
        grid[row][col] = 0;
        lands[0]++;
        for(int i=0; i< 4; i++){
            int Nrow = delRow[i] + row;
            int Ncol = delCol[i] + col;
            dfs(Nrow, Ncol, grid, lands);
        }
    }
    public int numEnclaves(int[][] grid) {
        int island = 0;
        int N = grid.length, M = grid[0].length;
        Queue<Pair> queue = new LinkedList<>();
        for(int i=0; i< N; i++){
            for(int j=0; j< M; j++){
                if(i == 0 || i == N-1 || j == 0 || j == M-1){
                    if(grid[i][j] == 1){
                        queue.add(new Pair(i, j));
                        grid[i][j] = 0;
                    }
                }
            }
        }
        int [] delRow = {-1, 0, 1, 0};
        int [] delCol = {0, 1, 0, -1};
        while(!queue.isEmpty()){
            Pair p = queue.remove();
            for(int i=0; i<4; i++){
                int nrow = delRow[i] + p.row;
                int ncol = delCol[i] + p.col;
                if(nrow >=0 && nrow < N && ncol >=0 && ncol < M && grid[nrow][ncol] == 1){
                    queue.add(new Pair(nrow, ncol));
                    grid[nrow][ncol] = 0;
                }
            }
        }
        for(int i=0; i< N; i++){
            for(int j=0; j< M; j++){
                System.out.print(grid[i][j] + " ");
            }
            System.out.println();
        }
        for(int i=0; i< N; i++){
            for(int j=0; j< M; j++){
                if(grid[i][j] == 1){
                    int [] lands = {0};
                    dfs(i, j, grid, lands);
                    island += lands[0];
                }
            }
        }
        return island;
    }
}
